File: SAMYPluginTemplates/SamyPlugins_Template_Cpp/src/Robot/robot.cpp
#include "robot.h"

Robot::Robot(std::string address_, Signals* signals_, Plugin* plugin_):
    address(address_),
    signals(signals_),
    plugin(plugin_)
    {
        //Set default values
        settings.transSpeedAbsolute.setting = 10;
        settings.transAccelAbsolute.setting = 10;
        settings.lengthUnitName = UA_CRCL_LENGTHUNITENUMDATATYPE_METER;
        settings.rotAccelAbsolute.setting = 1.2;
        settings.rotSpeedAbsolute.setting = 1.2;
    }

int Robot::InitCanon(UA_InitCanonParametersSetDataType* initCanon){
    spdlog::info("Robot: Execute InitCanon command");
    spdlog::info("InitCanon Name: {}", initCanon->name.data);
    return COMMAND_SUCCESS;
}

int Robot::EndCanon(UA_EndCanonParametersSetDataType* endCanon){
    spdlog::info("Robot: Execute EndCanon command");
    spdlog::info("EndCanon Name: {}", endCanon->name.data);
    return COMMAND_SUCCESS;
}

int Robot::Message(UA_MessageParametersSetDataType* message){
    spdlog::info("Robot: Execute Message command");
    spdlog::info("Message Name: {}", message->name.data);
    return COMMAND_SUCCESS;
}

int Robot::MoveTo(UA_MoveToParametersSetDataType* moveTo){
    spdlog::info("Robot: Execute MoveTo command");
    spdlog::info("MoveStraight {}", (bool) moveTo->moveStraight);
    spdlog::info("MoveTo Name: {}", (char*) moveTo->name.data);
    //Test write infoSource
    UA_CRCL_PoseDataType pose1;
    pose1.id = 743;
    pose1.name = UA_STRING( "POSE1" );
    pose1.point.id = 87;
    pose1.point.name = UA_STRING( "POINT NAME" );
    pose1.point.x = 0.987;
    pose1.point.y = 6.543;
    pose1.point.z = 1.467;
    pose1.xAxis.id = 8711;
    pose1.xAxis.name = UA_STRING( "xVector NAME" );
    pose1.xAxis.i = 0.987;
    pose1.xAxis.j = 6.543;
    pose1.xAxis.k = 1.467;
    pose1.zAxis.id = 8764;
    pose1.zAxis.name = UA_STRING( "zVector NAME" );
    pose1.zAxis.i = 5.987;
    pose1.zAxis.j = 8.543;
    pose1.zAxis.k = 121.467;
    UA_NodeId id = plugin->GetInformationSourceNodeId("SuperPose_0");
    bool ret = WriteInfoSource(&pose1, &UA_TYPES_CRCL[UA_TYPES_CRCL_CRCL_POSEDATATYPE],
                    id, plugin->samy_core_client_read);
    if (ret){
        return COMMAND_SUCCESS;
    } else {
        signals->Halt(); // Tells the SAMYCore that the execution of the skill failed by setting the
                         // state of the skill to halted.
        return COMMAND_ERROR;
    }
}

int Robot::MoveScrew(UA_MoveScrewParametersSetDataType* moveScrew){
    spdlog::info("Robot: Execute MoveScrew command");
    spdlog::info("MoveScrew Name: {}", moveScrew->name.data);
    return COMMAND_SUCCESS;
}

int Robot::MoveThroughTo(UA_MoveThroughToParametersSetDataType* moveThroughTo){
    spdlog::info("Robot: Execute MoveThroughTo command");
    spdlog::info("MoveThroughTo Name: {}", moveThroughTo->name.data);
    return COMMAND_SUCCESS;
}

int Robot::Dwell(UA_DwellParametersSetDataType* dwell){
    spdlog::info("Robot: Execute Dwell command");
    spdlog::info("Dwell Name: {}", dwell->name.data);
    return COMMAND_SUCCESS;
}

int Robot::ActuateJoints(UA_ActuateJointsParametersSetDataType* actuateJoints){
    spdlog::info("Robot: Execute ActuateJoints command");
    spdlog::info("ActuateJoints Name: {}", actuateJoints->name.data);
    return COMMAND_SUCCESS;
}

int Robot::ConfigureJointReports(UA_ConfigureJointReportsParametersSetDataType* configureJointReports){
    spdlog::info("Robot: Execute ConfigureJointReports command");
    spdlog::info("ConfigureJointReports Name: {}", configureJointReports->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetDefaultJointPositionsTolerances(UA_SetDefaultJointPositionsTolerancesParametersSetDataType* setDefaultJointPositionsTolerances){
    spdlog::info("Robot: Execute SetDefaultJointPositionsTolerances command");
    spdlog::info("SetDefaultJointPositionsTolerances Name: {}", setDefaultJointPositionsTolerances->name.data);
    return COMMAND_SUCCESS;
}

int Robot::GetStatus(UA_GetStatusParametersSetDataType* getStatus){
    spdlog::info("Robot: Execute GetStatus command");
    spdlog::info("GetStatus Name: {}", getStatus->name.data);
    return COMMAND_SUCCESS;
}

int Robot::CloseToolChanger(UA_CloseToolChangerParametersSetDataType* closeToolChanger){
    spdlog::info("Robot: Execute CloseToolChanger command");
    spdlog::info("CloseToolChanger Name: {}", closeToolChanger->name.data);
    return COMMAND_SUCCESS;
}

int Robot::OpenToolChanger(UA_OpenToolChangerParametersSetDataType* openToolChanger){
    spdlog::info("Robot: Execute OpenToolChanger command");
    spdlog::info("OpenToolChanger Name: {}", openToolChanger->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetRobotParameters(UA_SetRobotParametersParametersSetDataType* setRobotParameters){
    spdlog::info("Robot: Execute SetRobotParameters command");
    spdlog::info("SetRobotParameters Name: {}", setRobotParameters->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetEndeffectorParameters(UA_SetEndeffectorParametersParametersSetDataType* setEndeffectorParameters){
    spdlog::info("Robot: Execute SetEndeffectorParameters command");
    spdlog::info("SetEndeffectorParameters Name: {}", setEndeffectorParameters->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetEndeffector(UA_SetEndeffectorParametersSetDataType* setEndeffector){
    spdlog::info("Robot: Execute SetEndeffector command");
    spdlog::info("SetEndeffector Name: {}", setEndeffector->name.data);
    spdlog::info("SetEndeffector: Value {:f}",setEndeffector->setting.fraction);
    return COMMAND_SUCCESS;
}

int Robot::SetTransAccel(UA_SetTransAccelParametersSetDataType* setTransAccel){
    spdlog::info("Robot: Execute SetTransAccel command");
    spdlog::info("SetTransAccel Name: {}", setTransAccel->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetTransSpeed(UA_SetTransSpeedParametersSetDataType* setTransSpeed){
    spdlog::info("Robot: Execute SetTransSpeed command");
    spdlog::info("SetTransSpeed Name: {}", setTransSpeed->name.data);
    if (setTransSpeed->transSpeed.switchField == UA_CRCL_TRANSSPEEDDATATYPESWITCH_TRANSSPEEDABSOLUTEDATATYPE){
        settings.transSpeedAbsolute.setting = setTransSpeed->transSpeed.fields.transSpeedAbsoluteDataType.setting;
    } else {
        settings.transSpeedRelative.fraction.fraction = setTransSpeed->transSpeed.fields.transSpeedRelativeDataType.fraction.fraction;
    }
    return COMMAND_SUCCESS;
}

int Robot::SetRotAccel(UA_SetRotAccelParametersSetDataType* setRotAccel){
    spdlog::info("Robot: Execute SetRotAccel command");
    spdlog::info("SetRotAccel Name: {}", setRotAccel->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetRotSpeed(UA_SetRotSpeedParametersSetDataType* setRotSpeed){
    spdlog::info("Robot: Execute SetRotSpeed command");
    spdlog::info("SetRotSpeed Name: {}", setRotSpeed->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetAngleUnits(UA_SetAngleUnitsParametersSetDataType* setAngleUnits){
    spdlog::info("Robot: Execute SetAngleUnits command");
    spdlog::info("SetAngleUnits Name: {}", setAngleUnits->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetEndPoseTolerance(UA_SetEndPoseToleranceParametersSetDataType* setEndPoseTolerance){
    spdlog::info("Robot: Execute SetEndPoseTolerance command");
    spdlog::info("SetEndPoseTolerance Name: {}", setEndPoseTolerance->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetForceUnits(UA_SetForceUnitsParametersSetDataType* setForceUnits){
    spdlog::info("Robot: Execute SetForceUnits command");
    spdlog::info("SetForceUnits Name: {}", setForceUnits->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetIntermediatePoseTolerance(UA_SetIntermediatePoseToleranceParametersSetDataType* setIntermediatePoseTolerance){
    spdlog::info("Robot: Execute SetIntermediatePoseTolerance command");
    spdlog::info("SetIntermediatePoseTolerance Name: {}", setIntermediatePoseTolerance->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetLengthUnits(UA_SetLengthUnitsParametersSetDataType* setLengthUnits){
    spdlog::info("Robot: Execute SetLengthUnits command");
    spdlog::info("SetLengthUnits Name: {}", setLengthUnits->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetMotionCoordination(UA_SetMotionCoordinationParametersSetDataType* setMotionCoordination){
    spdlog::info("Robot: Execute SetMotionCoordination command");
    spdlog::info("SetMotionCoordination Name: {}", setMotionCoordination->name.data);
    return COMMAND_SUCCESS;
}

int Robot::SetTorqueUnits(UA_SetTorqueUnitsParametersSetDataType* setTorqueUnits){
    spdlog::info("Robot: Execute SetTorqueUnits command");
    spdlog::info("SetTorqueUnits Name: {}", setTorqueUnits->name.data);
    return COMMAND_SUCCESS;
}

int Robot::StopMotion(UA_StopMotionParametersSetDataType* stopMotion){
    spdlog::info("Robot: Execute StopMotion command");
    spdlog::info("StopMotion Name: {}", stopMotion->name.data);
    return COMMAND_SUCCESS;
}

int Robot::ConfigureStatusReport(UA_ConfigureStatusReportParametersSetDataType* configureStatusReport){
    spdlog::info("Robot: Execute ConfigureStatusReport command");
    spdlog::info("ConfigureStatusReport Name: {}", configureStatusReport->name.data);
    return COMMAND_SUCCESS;
}

int Robot::EnableSensor(UA_EnableSensorParametersSetDataType* enableSensor){
    spdlog::info("Robot: Execute EnableSensor command");
    spdlog::info("EnableSensor Name: {}", enableSensor->name.data);
    return COMMAND_SUCCESS;
}

int Robot::DisableSensor(UA_DisableSensorParametersSetDataType* disableSensor){
    spdlog::info("Robot: Execute DisableSensor command");
    spdlog::info("DisableSensor Name: {}", disableSensor->name.data);
    return COMMAND_SUCCESS;
}

int Robot::EnableGripper(UA_EnableGripperParametersSetDataType* enableGripper){
    spdlog::info("Robot: Execute EnableGripper command");
    spdlog::info("EnableGripper Name: {}", enableGripper->name.data);
    return COMMAND_SUCCESS;
}

int Robot::DisableGripper(UA_DisableGripperParametersSetDataType* disableGripper){
    spdlog::info("Robot: Execute DisableGripper command");
    spdlog::info("DisableGripper Name: {}", disableGripper->name.data);
    return COMMAND_SUCCESS;
}

int Robot::EnableRobotParameterStatus(UA_EnableRobotParameterStatusParametersSetDataType* enableRobotParameterStatus){
    spdlog::info("Robot: Execute EnableRobotParameterStatus command");
    spdlog::info("EnableRobotParameterStatus Name: {}", enableRobotParameterStatus->name.data);
    return COMMAND_SUCCESS;
}

int Robot::DisableRobotParameterStatus(UA_DisableRobotParameterStatusParametersSetDataType* disableRobotParameterStatus){
    spdlog::info("Robot: Execute DisableRobotParameterStatus command");
    spdlog::info("DisableRobotParameterStatus Name: {}", disableRobotParameterStatus->name.data);
    return COMMAND_SUCCESS;
}

File: SAMYPluginTemplates/SAMYPlugins_Template_Python/tests/robot/robot.py

import logging
import os
import sys
import time
PACKAGE_PARENT = '..'
SCRIPT_DIR = os.path.dirname(os.path.realpath(os.path.join(os.getcwd(), os.path.expanduser(__file__))))
sys.path.append(os.path.normpath(os.path.join(SCRIPT_DIR, PACKAGE_PARENT)))

from pubsub import pub
import samyplugin.CRCL_DataTypes
#from transform import Transform

class Robot(object):
    # This class represents the robot.

    def __init__(self, host):
        # create a logger for the robot object
        self.logger = logging.getLogger(__name__)
        self.logger.setLevel(logging.INFO)
        fh = logging.StreamHandler()
        fh.setFormatter(logging.Formatter("%(levelname)s %(filename)s - %(message)s"))
        self.logger.addHandler(fh)
        self.host = host

        # Subscribe to the CRCL command topics
        pub.subscribe(self.init_canon, "InitCanon")
        pub.subscribe(self.end_canon, "EndCanon")
        pub.subscribe(self.message, "Message")
        pub.subscribe(self.move_to, "MoveTo")
        pub.subscribe(self.move_screw, "MoveScrew")
        pub.subscribe(self.move_through_to, "MoveThroughTo")
        pub.subscribe(self.dwell, "Dwell")
        pub.subscribe(self.actuate_joints, "ActuateJoints")
        pub.subscribe(self.configure_joint_reports, "ConfigureJointReports")
        #pub.subscribe(self.configure_joint_report, "ConfigureJointReport")
        pub.subscribe(self.set_default_joint_positions_tolerances, "SetDefaultJointPositionsTolerances")
        pub.subscribe(self.get_status, "GetStatus")
        pub.subscribe(self.close_tool_changer, "CloseToolChanger")
        pub.subscribe(self.open_tool_changer, "OpenToolChanger")
        pub.subscribe(self.set_robot_parameters, "SetRobotParameters")
        pub.subscribe(self.set_endeffector_parameters, "SetEndeffectorParameters")
        pub.subscribe(self.set_endeffector, "SetEndeffector")
        pub.subscribe(self.set_trans_accel, "SetTransAccel")
        pub.subscribe(self.set_trans_speed, "SetTransSpeed")
        pub.subscribe(self.set_rot_accel, "SetRotAccel")
        pub.subscribe(self.set_rot_speed, "SetRotSpeed")
        pub.subscribe(self.set_angle_units, "SetAngleUnits")
        pub.subscribe(self.set_end_pose_tolerance, "SetEndPoseTolerance")
        pub.subscribe(self.set_force_units, "SetForceUnits")
        pub.subscribe(self.set_intermediate_pose_tolerance, "SetIntermediatePoseTolerance")
        pub.subscribe(self.set_length_units, "SetLengthUnits")
        pub.subscribe(self.set_motion_coordination, "SetMotionCoordination")
        pub.subscribe(self.set_torque_units, "SetTorqueUnits")
        pub.subscribe(self.stop_motion, "StopMotion")
        pub.subscribe(self.configure_status_report, "ConfigureStatusReport")
        pub.subscribe(self.enable_sensor, "EnableSensor")
        pub.subscribe(self.enable_gripper, "EnableGripper")
        pub.subscribe(self.disable_gripper, "DisableGripper")
        pub.subscribe(self.enable_robot_parameter_status, "EnableRobotParameterStatus")
        pub.subscribe(self.disable_sensor, "DisableSensor")
        pub.subscribe(self.disable_robot_parameter_status, "DisableRobotParameterStatus")


    # Add robot specific methods here:

    def init_canon(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="InitCanon", data=data)

    def end_canon(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="EndCanon", data=data)

    def message(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="Message", data=data)

    def move_to(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="MoveTo", data=data)

    def move_screw(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="MoveScrew", data=data)

    def move_through_to(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="MoveThroughTo", data=data)

    def dwell(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="Dwell", data=data)

    def actuate_joints(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="ActuateJoints", data=data)

    def configure_joint_reports(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="ConfigureJointReports", data=data)

    def configure_joint_report(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="ConfigureJointReport", data=data)

    def set_default_joint_positions_tolerances(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetDefaultJointPositionsTolerances", data=data)

    def get_status(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="GetStatus", data=data)

    def close_tool_changer(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="CloseToolChanger", data=data)

    def open_tool_changer(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="OpenToolChanger", data=data)

    def set_robot_parameters(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetRobotParameters", data=data)

    def set_endeffector_parameters(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetEndeffectorParameters", data=data)

    def set_endeffector(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetEndeffector", data=data)

    def set_trans_accel(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetTransAccel", data=data)

    def set_trans_speed(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetTransSpeed", data=data)

    def set_rot_accel(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetRotAccel", data=data)

    def set_rot_speed(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetRotSpeed", data=data)

    def set_angle_units(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetAngleUnits", data=data)

    def set_end_pose_tolerance(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetEndPoseTolerance", data=data)

    def set_force_units(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetForceUnits", data=data)

    def set_intermediate_pose_tolerance(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetIntermediatePoseTolerance", data=data)

    def set_length_units(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetLengthUnits", data=data)

    def set_motion_coordination(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetMotionCoordination", data=data)

    def set_torque_units(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="SetTorqueUnits", data=data)

    def stop_motion(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="StopMotion", data=data)

    def configure_status_report(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="ConfigureStatusReport", data=data)

    def enable_sensor(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="EnableSensor", data=data)

    def enable_gripper(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="EnableGripper", data=data)

    def disable_gripper(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="DisableGripper", data=data)

    def enable_robot_parameter_status(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="EnableRobotParameterStatus", data=data)

    def disable_sensor(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="DisableSensor", data=data)

    def disable_robot_parameter_status(self, data):
        self.logger.info("Sending data to info Source")
        pub.sendMessage("write_information_source", name="DisableRobotParameterStatus", data=data)
